presto-main/src/main/java/com/facebook/presto/execution/FutureStateChange.java

```diff
@@ -66,14 +69,54 @@ public void complete(T newState, Executor executor)
     private void fireStateChange(T newState, Executor executor)
     {
         requireNonNull(executor, "executor is null");
-        Set<SettableFuture<T>> futures;
+        Set<StateTrackingFuture<T>> futures;
         synchronized (listeners) {
             futures = ImmutableSet.copyOf(listeners);
             listeners.clear();
         }
 
-        for (SettableFuture<T> future : futures) {
+        for (StateTrackingFuture<T> future : futures) {
             executor.execute(() -> future.set(newState));
         }
     }
+
+    /**
+     * Future cancellation can be expensive, because when a future is canceled
+     * a CancellationException is created and thrown under the hood. To get rid
+     * of the cancellation overhead this future implementation just cancels the
+     * future with the same exception instance.
+     */
+    private static final class StateTrackingFuture<V>
+            extends AbstractFuture<V>
+    {
+        private static final CancellationException EXCEPTION = new CancellationException("Future is canceled");
+        private AtomicBoolean canceled = new AtomicBoolean();
+
+        public static <V> StateTrackingFuture<V> create()
+        {
+            return new StateTrackingFuture<V>();
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning)
+        {
+            if (canceled.compareAndSet(false, true)) {
+                setException(EXCEPTION);
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        protected boolean set(@Nullable V value)
```

Comment:
we can  only make it `protected` or `public` as the super class has it as `protected` (we cannot restrict further), so I made it `protected`.